### DISCUSSION TITLE
How to restart PoolController and Dash server - w/ out server reboot

### DISCUSSION THREAD

**Discussion by sdpools** [tagyoureit/nodejs-poolController#408] [App and O/S Config Questions] [answered]
Hello,

Everytime I have an issue with poolcontroler or dash, I am restarting the pi. If I try npm start& again it says the port is already in use. Seems like the network port bindings stay open after a crash/ halt. So, I just reboot. But this is added a ton of time to TS.

I have added the use of "screen" to create ssh sessions to resume. This doesn't help with restarting the poolcontroller apps.

Any suggestions?

**Comment by rstrouse** [comment]
If you are using pm2 just do `pm2 restart njsPC` and `pm2 restart dashPanel`.  The network port bindings are not left open after a crash.  The instance is likely still running somewhere but the node process has it halted.

**Reply by sdpools** [reply]
Rstrouse,

I hadn’t started using pm2 becuase I didn’t want the program to launch automatically on startup and not be able to monitor it. Once the config was finished I was going to add it.

Thank you for sharing this, I will use it after config is complete.

**Comment by rstrouse** [comment]
You should consider running the pi headless using Remote Desktop or VNC.  Then you have access to the entire pi without all the ssh pain and no monitors involved.

**Reply by sdpools** [reply]
Thank you for this suggestion - This was what I was looking for. A way to walk away for any amount of time and return to the cli for debug / troubleshooting.

Upon using pi vnc viewer, the speed was so bad that I couldn’t bear it. It seemed to have obvious read issues. I had used the OS when I first bought the pi4 and didn’t remember it being that bad.

Then it hit me - I began wondering if my microSD was giving up. 

It was an off brand - Patriot - running pi server tasks for 2+yrs.

This may also explain why my poolConfig.json PUT commands are not writing correctly. Debug shows it tries 3 times and gives up - never actually saving. Eventually, after the three attempts the server quits trying.

It also may explain why pool controller only runs for a few hours before halting / freezing.

I have ordered a replacement mSD (Sandisk extreme pro 32GB) and will report back.

**Reply by sdpools** [reply]
Update. I received the new Scandisc EVO mSD. I copied the imagine over (super fast) and restarted the server. 

Pi OS / CLI is noticeable faster!!

Unfortunately, it didn't correct the PUT command writing issue I am experiencing.

What did.

Out of frustration; I reset the IT i7 unit again via the OCP front panel.

Immediately after, I began programming and they stuck.

Wahoo.

I have noticed this before. I get these windows when everything works as it should,but never long enough to finish the whole config.

So if your having trouble writing and willing to erase the current setup- try a reset on the OCP.

I also remember reading about a "SP.zip" file that corrected the usb 485 comm line. If I continue to have issues I will look that up again and try it.

**Comment by tagyoureit** [comment, marked as answer]
You can find directions for [PM2 in the wiki](https://github.com/tagyoureit/nodejs-poolController/wiki/Automatically-start-at-boot---PM2-&-Systemd).  Running commands with the `&` will kill the app if an ssh session is terminated.

However, if you are want to see what node processes are running you can use `ps -ef | grep node`.  If you want to kill a rogue process you can use `sudo kill -9 [process_id]` but if things go south for me I just usually kill everything with `sudo killall node`.

**Reply by sdpools** [reply]
Thanks again. I thought pool controller was dying every time my ssh session closed. Good to know why. 

I have also switched to running vnc after the earlier suggestion and then starting the program from gui terminal. Allowing me to return later and check on it for debug.

I will stop using the & sign, as I don’t need that function (terminate w/ ssh).

I really appreciate the help.
